src/functions/projectConfig.ts

```diff
@@ -136,3 +221,39 @@ export function configForCodebase(config: ValidatedConfig, codebase: string): Va
   }
   return codebaseCfg;
 }
+
+/** Returns true if the codebase uses a local source. */
+export function isLocalConfig(c: ValidatedSingle): c is ValidatedLocalSingle {
+  return (c as ValidatedLocalSingle).source !== undefined;
+}
+
+/** Returns true if the codebase uses a remote source. */
+export function isRemoteConfig(c: ValidatedSingle): c is ValidatedRemoteSingle {
+  return (c as ValidatedRemoteSingle).remoteSource !== undefined;
+}
+
+/**
+ * Require a local functions config. Throws a FirebaseError if the config is remote.
+ * @param c The validated functions config entry.
+ * @param purpose Optional message to use in the error.
+ */
+export function requireLocal(c: ValidatedSingle, purpose?: string): ValidatedLocalSingle {
+  if (!isLocalConfig(c)) {
+    const msg =
+      purpose ??
+      "This operation requires a local functions source directory, but the codebase is configured with a remote source.";
+    throw new FirebaseError(msg);
+  }
+  return c;
+}
+
+/**
+ * Resolve the directory used for .env files.
+ * - Local: returns `configDir` if set, otherwise `source`.
+ * - Remote: returns `configDir` if set, otherwise `undefined`.
+ */
+export function resolveConfigDir(c: ValidatedSingle): string | undefined {
+  if (c.configDir) return c.configDir;
```

Comment:
nit: why not `return c.configDir || c.source`? Why do you care about isLocalConfig which is just a check for c.source anyway?